pageserver/src/walredo.rs

```diff
@@ -681,30 +684,82 @@ impl PostgresRedoProcess {
             })?;
 
         info!(
+            pid = child.id(),
             "launched WAL redo postgres process on {}",
             datadir.display()
         );
 
+        // NB: don't return an Err() from here on without kill()ing and wait()ing
+        //     for the child. Once we return it inside the Ok(...), it's the callers
+        //     responsibility to call Self::kill().
+
         let stdin = child.stdin.take().unwrap();
         let stdout = child.stdout.take().unwrap();
         let stderr = child.stderr.take().unwrap();
 
-        set_nonblock(stdin.as_raw_fd())?;
-        set_nonblock(stdout.as_raw_fd())?;
-        set_nonblock(stderr.as_raw_fd())?;
+        set_nonblock(stdin.as_raw_fd()).unwrap();
+        set_nonblock(stdout.as_raw_fd()).unwrap();
+        set_nonblock(stderr.as_raw_fd()).unwrap();
 
         Ok(PostgresRedoProcess {
+            tenant_id,
             child,
             stdin,
             stdout,
             stderr,
+            called_kill: false,
         })
     }
 
+    #[instrument(skip_all, fields(tenant_id=%self.tenant_id, pid=%self.child.id()))]
     fn kill(mut self) {
-        let _ = self.child.kill();
-        if let Ok(exit_status) = self.child.wait() {
-            error!("wal-redo-postgres exited with code {}", exit_status);
+        info!("killing wal-redo-postgres process");
+        self.called_kill = true;
+
+        // If false, we will panic upon errors in this function.
+        // If there are any wal-redo processes, they will be cleaned up like so:
+        // 1. Our write-side pipe FDs will be closed by kernel.
+        // 2. In response, walredo process will eventually hit SIGPIPE, either
+        //    when writing a result to stdout or reading a new command from stdin.
+        // 3. walredo process will exit and the supervisor process (systemd) will reap it.
+        let allow_leaks = cfg!(feature = "testing");
+
```

Comment:
We can think about this when we get to the actual fix.
Right now, I don't want to change the behavior except logging, so that we can confirm the root cause.

In general, I don't think it matters because we send a SIGKILL before .wait()ing. What tokio does is to avoid stupid mistakes like

```
process = spawn process with piped stdin
process.stdin.as_mut_ref().write(...)
# forget to close process.stdin
process.wait()
```

If the child process is a simple loop that reads stdin until EOF, this will cause the parent to wait forever.
Tokio papers over the error by closing stdin.

Reply:
Ack, just recording that I thought it was SIGTERM instead but we checked it should be SIGKILL. Also, luckily we don't target windows where I've ran into this deadlock :)